### PR TITLE
Refactors embedding as an element, adds toy throwing stars and sticky tape!

### PR DESCRIPTION
**Pull request by Ryll-Ryll**
## About The Pull Request
This PR moves as much embedding code out of human/carbon code and into a central element handler as I could. Originally it was gonna be a component, but then I realized the state variables could just be left in the weapon or target, and figured it might make a good case for an element! Honestly I didn't really plan this out and maybe this would have worked better as a subsystem, but oh well! Here we are!

Also adds new functionality for embeds: jostling, and stamina share. Every time a mob with an embed moves, they have a chance to jostle the embedded object and cause pain! I envision this being used with the other new feature, embed stamina share, which diverts a percentage of any embedding damage from brute to stamina. For example, I include a "shock throwing star" that has a 15% jostle rate but deals 80% of its damage to stamina, which would be useful for security in chases. Unlike weapons like disablers, stunbatons, and bolas, the shock throwing star is more useful the more the target tries to run rather than immediately, adding extra layers of depth.

ALSO in more fun news, adds support for harmless embeds! Adds a toy throwing star that does no damage and is fluffed as sticking _to_ the target, rather than _in_ them.

[](https://i.imgur.com/EgmThqI.png)

## Why It's Good For The Game
Centralizes most embedding code without exploding the memory footprint of embedding. Adds new functionality to embedding.

## Changelog
:cl: Ryll/Shaps
add: Adds harmless toy throwing stars that you can throw at the HoS and get killed for as if they were real! Fun for all!
/:cl:

### CODE REVIEW

File: code/game/objects/items.dm [hunk 1]
```diff
@@ -85,7 +85,19 @@ GLOBAL_VAR_INIT(rpg_loot_items, FALSE)
...
+	var/embedded_fall_chance = EMBEDDED_ITEM_FALLOUT
+	var/embedded_pain_chance = EMBEDDED_PAIN_CHANCE
+	var/embedded_pain_multiplier = EMBEDDED_PAIN_MULTIPLIER //The coefficient of multiplication for the damage this item does while embedded (this*w_class)
+	var/embedded_fall_pain_multiplier = EMBEDDED_FALL_PAIN_MULTIPLIER //The coefficient of multiplication for the damage this item does when falling out of a limb (this*w_class)
+	var/embedded_impact_pain_multiplier = EMBEDDED_IMPACT_PAIN_MULTIPLIER  //The coefficient of multiplication for the damage this item does when first embedded (this*w_class)
+	var/embedded_unsafe_removal_pain_multiplier = EMBEDDED_UNSAFE_REMOVAL_PAIN_MULTIPLIER //The coefficient of multiplication for the damage removing this without surgery causes (this*w_class)
+	var/embedded_unsafe_removal_time = EMBEDDED_UNSAFE_REMOVAL_TIME //A time in ticks, multiplied by the w_class.
+	var/embedded_ignore_throwspeed_threshold = FALSE //if we don't give a damn about EMBED_THROWSPEED_THRESHOLD
+
+	var/embedded_in
```
Comment (Qustinnus) [inline code/game/objects/items.dm]: ah yes the types of things you would find on a component! If only we had such magic things.

File: code/_compile_options.dm [hunk 2]
```diff
@@ -1,5 +1,5 @@
-//#define TESTING				//By using the testing("message") proc you can create debug-feedback for people with this
-								//uncommented, but not visible in the release version)
+#define TESTING				//By using the testing("message") proc you can create debug-feedback for people with this
+#define LOWMEMORYMODE								//uncommented, but not visible in the release version)
```
Comment (PKPenguin321) [inline code/_compile_options.dm]: 👀

Comment (Shapsy) [reply code/_compile_options.dm]: sh

**Comment by Shapsy** [comment]
Added sticky tape, an item that turns any object that isn't already embeddable to a harmless embeddable!

[](https://i.imgur.com/Tz3pHrd.png)

File: code/datums/elements/embed.dm [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+	if(!isitem(target))
+		return ELEMENT_INCOMPATIBLE
+
+	if(!target.embedding)
+		target.embedding = getEmbeddingBehavior() // if it couldn't embed before, it can now (pending throwspeed, anyway)!
+
+	RegisterSignal(target, COMSIG_MOVABLE_IMPACT_ZONE, .proc/embed_check)
+	RegisterSignal(target, COMSIG_ITEM_EMBED_REMOVING_RIP, .proc/rip_out)
+	RegisterSignal(target, COMSIG_ITEM_EMBED_REMOVE_SURGERY, .proc/safe_remove)
+	RegisterSignal(target, COMSIG_PARENT_QDELETING, .proc/byeee)
```
Comment (kriskog) [inline code/datums/elements/embed.dm]: This makes Travis sad. This signal is already listened for in this element, registered in the parent proc. Add the proc call to byeee to Detached or add the contents there.

File: code/datums/elements/embed.dm [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+/datum/element/embed/Detach(target)
+	. = ..()
+	UnregisterSignal(target, COMSIG_MOVABLE_IMPACT_ZONE)
+	UnregisterSignal(target, COMSIG_ITEM_EMBED_REMOVING_RIP)
+	UnregisterSignal(target, COMSIG_ITEM_EMBED_REMOVE_SURGERY)
+	UnregisterSignal(target, COMSIG_PARENT_QDELETING)
```
Comment (kriskog) [inline code/datums/elements/embed.dm]: This is handled in parent

File: code/datums/elements/embed.dm [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+	if(!target.embedding)
+		target.embedding = getEmbeddingBehavior() // if it couldn't embed before, it can now (pending throwspeed, anyway)!
+
+	RegisterSignal(target, COMSIG_MOVABLE_IMPACT_ZONE, .proc/embed_check)
+	RegisterSignal(target, COMSIG_ITEM_EMBED_REMOVING_RIP, .proc/rip_out)
+	RegisterSignal(target, COMSIG_ITEM_EMBED_REMOVE_SURGERY, .proc/safe_remove)
+	RegisterSignal(target, COMSIG_PARENT_QDELETING, .proc/byeee)
+
+/datum/element/embed/Detach(target)
```
Comment (kriskog) [inline code/datums/elements/embed.dm]: Update the parameters to match the parent proc. Remember to update the vars used in unregister below.

**Review by kriskog — CHANGES_REQUESTED** [review]
Travis angry

**Comment by kriskog** [comment]
You messed up the merge here. 
image

**Review by stylemistake — CHANGES_REQUESTED** [review]
This clearly needs some work.

**Comment by Shapsy** [comment]
okay I fell asleep before I got around to fixing it...

**Comment by Shapsy** [comment]
putting this on ice

**Pull request closed without merging**